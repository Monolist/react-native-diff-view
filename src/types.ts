import { IHunk, IChange } from 'gitdiff-parser';
import { TextStyle, ViewStyle } from 'react-native';

export type DiffType = 'add' | 'delete' | 'modify' | 'rename' | 'copy';

export type Side = 'old' | 'new';

export type ViewType = 'unified' | 'split';

export type GutterType = 'default' | 'none' | 'anchor';

export type Widgets = Record<string, JSX.Element>;

export type Events = Record<string, (arg: any) => any>;

export type DiffChildren = (hunks: IHunk[]) => JSX.Element | JSX.Element[];

export interface IGutterOptions {
  change: IChange;
  side: Side;
  inHoverState: boolean;
  textStyle?: TextStyle;
  renderDefault: (style?: TextStyle, textStyle?: ViewStyle) => JSX.Element;
}
