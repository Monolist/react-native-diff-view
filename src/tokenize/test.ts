import refractor from 'refractor';
import { IHunk } from '../types';
import { basicHunk, multipleHunk } from '../fixtures';
import toTokenTrees from './to_token_trees';
import { markEdits, markWord, tokenize, ITokenizeOptions } from '.';

describe('tokenize', () => {
  describe('markEdits', () => {
    it('returns a function', () => {
      expect(typeof markEdits([], {})).toBe('function');
    });

    it('empty', () => {
      expect(markEdits([], {})([[{} as any], [{}]])).toEqual([[{}], [{}]]);
      expect(markEdits([], { type: 'line' })([[{} as any], [{}]])).toEqual([[{}], [{}]]);
    });

    it('withHunk', () => {
      expect(markEdits([basicHunk], {})([[], []])).toMatchSnapshot();
      expect(markEdits([multipleHunk], {})([[], []])).toMatchSnapshot();
      expect(markEdits([basicHunk], { type: 'line' })([[], []])).toMatchSnapshot();
      expect(markEdits([multipleHunk], { type: 'line' })([[], []])).toMatchSnapshot();
    });
  });

  describe('markWord', () => {
    it('returns a function', () => {
      expect(typeof markWord()).toBe('function');
    });

    it('empty', () => {
      expect(markWord()([[{} as any], []])).toEqual([[[]], []]);
    });

    it('hit value', () => {
      expect(
        markWord('A')([
          [[[{ value: 'A' }]]], // oldLinesOfPaths
          [[[{ value: 'A' }]]], // newLinesOfPaths
        ]),
      ).toEqual([
        [[[{ markType: undefined, type: 'mark', value: 'A' }]]],
        [[[{ markType: undefined, type: 'mark', value: 'A' }]]],
      ]);
    });

    it('mark single word', () => {
      expect(
        markWord('A', 'first', 'a')([
          [[[{ value: 'AAaabb' }]]], // oldLinesOfPaths
          [[[{ value: '' }]]], // newLinesOfPaths
        ]),
      ).toEqual([
        [
          [
            [
              {
                markType: 'first',
                type: 'mark',
                value: 'a',
              },
            ],
            [
              {
                markType: 'first',
                type: 'mark',
                value: 'a',
              },
            ],
            [
              {
                value: 'aabb',
              },
            ],
          ],
        ],
        [[[{ value: '' }]]],
      ]);
    });

    it('mark complex word', () => {
      expect(
        markWord('\t', 'tab', '    ')([
          [[[{ value: '\t\t    bb' }]]], // oldLinesOfPaths
          [[[{ value: '' }]]], // newLinesOfPaths
        ]),
      ).toEqual([
        [
          [
            [
              {
                markType: 'tab',
                type: 'mark',
                value: '    ',
              },
            ],
            [
              {
                markType: 'tab',
                type: 'mark',
                value: '    ',
              },
            ],
            [
              {
                value: '    bb',
              },
            ],
          ],
        ],
        [[[{ value: '' }]]],
      ]);
    });
  });

  describe('toTokenTrees', () => {
    const content =
      '<img style="width:<?php echo (80 / count($images)) ?>%" src=\'<?php echo Component_Union_FinanceInfoModel::UNION_LICEPIC_URL_PREFIX . $image?>\'/>';

    it('php will render __PHP__', () => {
      const options: ITokenizeOptions = {
        highlight: true,
        refractor,
        language: 'php',
        oldSource: null,
        enhancers: [],
      };

      const hunks = [
        {
          changes: [
            {
              content,
              isInsert: true,
              lineNumber: 1,
              type: 'insert',
            },
          ],
          content: '@@ -1,10 +1,17 @@',
          isPlain: false,
          newLines: 17,
          newStart: 1,
          oldLines: 10,
          oldStart: 1,
        },
      ] as IHunk[];

      const tokens = toTokenTrees(hunks, options);

      expect(tokens).toMatchSnapshot();
    });

    it('refractor highlight', () => {
      expect(refractor.highlight(content, 'php')).toMatchSnapshot();
    });
  });

  describe('tokenize', () => {
    it('basic', () => {
      expect(tokenize([], {})).toEqual({
        new: [[{ type: 'text', value: '' }]],
        old: [[{ type: 'text', value: '' }]],
      });
    });

    it('different config', () => {
      expect(
        tokenize([], {
          highlight: true,
          refractor: { highlight: (text: string) => [{ type: 'text', value: text }] },
        }),
      ).toEqual({
        new: [[{ type: 'text', value: '' }]],
        old: [[{ type: 'text', value: '' }]],
      });

      expect(tokenize([], { oldSource: 'A' })).toEqual({
        new: [[{ type: 'text', value: 'A' }]],
        old: [[{ type: 'text', value: 'A' }]],
      });
    });

    it('withHunk', () => {
      expect(tokenize([basicHunk, multipleHunk], {})).toMatchSnapshot();
    });
  });
});
