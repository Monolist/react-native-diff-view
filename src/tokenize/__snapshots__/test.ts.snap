// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tokenize markEdits withHunk 1`] = `
Array [
  Array [],
  Array [],
]
`;

exports[`tokenize markEdits withHunk 2`] = `
Array [
  Array [],
  Array [],
]
`;

exports[`tokenize markEdits withHunk 3`] = `
Array [
  Array [],
  Array [],
]
`;

exports[`tokenize markEdits withHunk 4`] = `
Array [
  Array [],
  Array [],
]
`;

exports[`tokenize toTokenTrees php will render __PHP__ 1`] = `
Array [
  Object {
    "children": Array [],
    "type": "root",
  },
  Object {
    "children": Array [
      Object {
        "children": Array [
          Object {
            "children": Array [
              Object {
                "children": Array [
                  Object {
                    "type": "text",
                    "value": "<",
                  },
                ],
                "properties": Object {
                  "className": Array [
                    "token",
                    "punctuation",
                  ],
                },
                "tagName": "span",
                "type": "element",
              },
              Object {
                "type": "text",
                "value": "img",
              },
            ],
            "properties": Object {
              "className": Array [
                "token",
                "tag",
              ],
            },
            "tagName": "span",
            "type": "element",
          },
          Object {
            "children": Array [
              Object {
                "children": Array [
                  Object {
                    "type": "text",
                    "value": " ",
                  },
                  Object {
                    "children": Array [
                      Object {
                        "type": "text",
                        "value": "style",
                      },
                    ],
                    "properties": Object {
                      "className": Array [
                        "token",
                        "attr-name",
                      ],
                    },
                    "tagName": "span",
                    "type": "element",
                  },
                ],
                "properties": Object {
                  "className": Array [
                    "token",
                    "attr-name",
                  ],
                },
                "tagName": "span",
                "type": "element",
              },
              Object {
                "children": Array [
                  Object {
                    "type": "text",
                    "value": "=\\"",
                  },
                ],
                "properties": Object {
                  "className": Array [
                    "token",
                    "punctuation",
                  ],
                },
                "tagName": "span",
                "type": "element",
              },
              Object {
                "children": Array [
                  Object {
                    "children": Array [
                      Object {
                        "type": "text",
                        "value": "width",
                      },
                    ],
                    "properties": Object {
                      "className": Array [
                        "token",
                        "property",
                      ],
                    },
                    "tagName": "span",
                    "type": "element",
                  },
                  Object {
                    "children": Array [
                      Object {
                        "type": "text",
                        "value": ":",
                      },
                    ],
                    "properties": Object {
                      "className": Array [
                        "token",
                        "punctuation",
                      ],
                    },
                    "tagName": "span",
                    "type": "element",
                  },
                  Object {
                    "type": "text",
                    "value": "___PHP",
                  },
                  Object {
                    "children": Array [
                      Object {
                        "type": "text",
                        "value": "0",
                      },
                    ],
                    "properties": Object {
                      "className": Array [
                        "token",
                        "number",
                      ],
                    },
                    "tagName": "span",
                    "type": "element",
                  },
                  Object {
                    "type": "text",
                    "value": "___%",
                  },
                ],
                "properties": Object {
                  "className": Array [
                    "token",
                    "attr-value",
                  ],
                },
                "tagName": "span",
                "type": "element",
              },
              Object {
                "children": Array [
                  Object {
                    "type": "text",
                    "value": "\\"",
                  },
                ],
                "properties": Object {
                  "className": Array [
                    "token",
                    "punctuation",
                  ],
                },
                "tagName": "span",
                "type": "element",
              },
            ],
            "properties": Object {
              "className": Array [
                "token",
                "style-attr",
                "language-css",
              ],
            },
            "tagName": "span",
            "type": "element",
          },
          Object {
            "type": "text",
            "value": " ",
          },
          Object {
            "children": Array [
              Object {
                "type": "text",
                "value": "src",
              },
            ],
            "properties": Object {
              "className": Array [
                "token",
                "attr-name",
              ],
            },
            "tagName": "span",
            "type": "element",
          },
          Object {
            "children": Array [
              Object {
                "children": Array [
                  Object {
                    "type": "text",
                    "value": "=",
                  },
                ],
                "properties": Object {
                  "className": Array [
                    "token",
                    "punctuation",
                  ],
                },
                "tagName": "span",
                "type": "element",
              },
              Object {
                "children": Array [
                  Object {
                    "type": "text",
                    "value": "'",
                  },
                ],
                "properties": Object {
                  "className": Array [
                    "token",
                    "punctuation",
                  ],
                },
                "tagName": "span",
                "type": "element",
              },
              Object {
                "type": "text",
                "value": "___PHP1___",
              },
              Object {
                "children": Array [
                  Object {
                    "type": "text",
                    "value": "'",
                  },
                ],
                "properties": Object {
                  "className": Array [
                    "token",
                    "punctuation",
                  ],
                },
                "tagName": "span",
                "type": "element",
              },
            ],
            "properties": Object {
              "className": Array [
                "token",
                "attr-value",
              ],
            },
            "tagName": "span",
            "type": "element",
          },
          Object {
            "children": Array [
              Object {
                "type": "text",
                "value": "/>",
              },
            ],
            "properties": Object {
              "className": Array [
                "token",
                "punctuation",
              ],
            },
            "tagName": "span",
            "type": "element",
          },
        ],
        "properties": Object {
          "className": Array [
            "token",
            "tag",
          ],
        },
        "tagName": "span",
        "type": "element",
      },
    ],
    "type": "root",
  },
]
`;

exports[`tokenize toTokenTrees refractor highlight 1`] = `
Array [
  Object {
    "children": Array [
      Object {
        "children": Array [
          Object {
            "children": Array [
              Object {
                "type": "text",
                "value": "<",
              },
            ],
            "properties": Object {
              "className": Array [
                "token",
                "punctuation",
              ],
            },
            "tagName": "span",
            "type": "element",
          },
          Object {
            "type": "text",
            "value": "img",
          },
        ],
        "properties": Object {
          "className": Array [
            "token",
            "tag",
          ],
        },
        "tagName": "span",
        "type": "element",
      },
      Object {
        "children": Array [
          Object {
            "children": Array [
              Object {
                "type": "text",
                "value": " ",
              },
              Object {
                "children": Array [
                  Object {
                    "type": "text",
                    "value": "style",
                  },
                ],
                "properties": Object {
                  "className": Array [
                    "token",
                    "attr-name",
                  ],
                },
                "tagName": "span",
                "type": "element",
              },
            ],
            "properties": Object {
              "className": Array [
                "token",
                "attr-name",
              ],
            },
            "tagName": "span",
            "type": "element",
          },
          Object {
            "children": Array [
              Object {
                "type": "text",
                "value": "=\\"",
              },
            ],
            "properties": Object {
              "className": Array [
                "token",
                "punctuation",
              ],
            },
            "tagName": "span",
            "type": "element",
          },
          Object {
            "children": Array [
              Object {
                "children": Array [
                  Object {
                    "type": "text",
                    "value": "width",
                  },
                ],
                "properties": Object {
                  "className": Array [
                    "token",
                    "property",
                  ],
                },
                "tagName": "span",
                "type": "element",
              },
              Object {
                "children": Array [
                  Object {
                    "type": "text",
                    "value": ":",
                  },
                ],
                "properties": Object {
                  "className": Array [
                    "token",
                    "punctuation",
                  ],
                },
                "tagName": "span",
                "type": "element",
              },
              Object {
                "type": "text",
                "value": "___PHP",
              },
              Object {
                "children": Array [
                  Object {
                    "type": "text",
                    "value": "0",
                  },
                ],
                "properties": Object {
                  "className": Array [
                    "token",
                    "number",
                  ],
                },
                "tagName": "span",
                "type": "element",
              },
              Object {
                "type": "text",
                "value": "___%",
              },
            ],
            "properties": Object {
              "className": Array [
                "token",
                "attr-value",
              ],
            },
            "tagName": "span",
            "type": "element",
          },
          Object {
            "children": Array [
              Object {
                "type": "text",
                "value": "\\"",
              },
            ],
            "properties": Object {
              "className": Array [
                "token",
                "punctuation",
              ],
            },
            "tagName": "span",
            "type": "element",
          },
        ],
        "properties": Object {
          "className": Array [
            "token",
            "style-attr",
            "language-css",
          ],
        },
        "tagName": "span",
        "type": "element",
      },
      Object {
        "type": "text",
        "value": " ",
      },
      Object {
        "children": Array [
          Object {
            "type": "text",
            "value": "src",
          },
        ],
        "properties": Object {
          "className": Array [
            "token",
            "attr-name",
          ],
        },
        "tagName": "span",
        "type": "element",
      },
      Object {
        "children": Array [
          Object {
            "children": Array [
              Object {
                "type": "text",
                "value": "=",
              },
            ],
            "properties": Object {
              "className": Array [
                "token",
                "punctuation",
              ],
            },
            "tagName": "span",
            "type": "element",
          },
          Object {
            "children": Array [
              Object {
                "type": "text",
                "value": "'",
              },
            ],
            "properties": Object {
              "className": Array [
                "token",
                "punctuation",
              ],
            },
            "tagName": "span",
            "type": "element",
          },
          Object {
            "type": "text",
            "value": "___PHP1___",
          },
          Object {
            "children": Array [
              Object {
                "type": "text",
                "value": "'",
              },
            ],
            "properties": Object {
              "className": Array [
                "token",
                "punctuation",
              ],
            },
            "tagName": "span",
            "type": "element",
          },
        ],
        "properties": Object {
          "className": Array [
            "token",
            "attr-value",
          ],
        },
        "tagName": "span",
        "type": "element",
      },
      Object {
        "children": Array [
          Object {
            "type": "text",
            "value": "/>",
          },
        ],
        "properties": Object {
          "className": Array [
            "token",
            "punctuation",
          ],
        },
        "tagName": "span",
        "type": "element",
      },
    ],
    "properties": Object {
      "className": Array [
        "token",
        "tag",
      ],
    },
    "tagName": "span",
    "type": "element",
  },
]
`;

exports[`tokenize tokenize withHunk 1`] = `
Object {
  "new": Array [
    Array [
      Object {
        "type": "text",
        "value": "",
      },
    ],
    Array [
      Object {
        "type": "text",
        "value": "",
      },
    ],
    Array [
      Object {
        "type": "text",
        "value": "",
      },
    ],
    Array [
      Object {
        "type": "text",
        "value": "",
      },
    ],
    Array [
      Object {
        "type": "text",
        "value": "",
      },
    ],
    Array [
      Object {
        "type": "text",
        "value": "",
      },
    ],
    Array [
      Object {
        "type": "text",
        "value": "",
      },
    ],
    Array [
      Object {
        "type": "text",
        "value": "",
      },
    ],
    Array [
      Object {
        "type": "text",
        "value": "",
      },
    ],
    Array [
      Object {
        "type": "text",
        "value": "",
      },
    ],
    Array [
      Object {
        "type": "text",
        "value": "",
      },
    ],
    Array [
      Object {
        "type": "text",
        "value": "",
      },
    ],
    Array [
      Object {
        "type": "text",
        "value": "",
      },
    ],
    Array [
      Object {
        "type": "text",
        "value": "",
      },
    ],
    Array [
      Object {
        "type": "text",
        "value": "",
      },
    ],
    Array [
      Object {
        "type": "text",
        "value": "",
      },
    ],
    Array [
      Object {
        "type": "text",
        "value": "",
      },
    ],
    Array [
      Object {
        "type": "text",
        "value": "",
      },
    ],
    Array [
      Object {
        "type": "text",
        "value": "",
      },
    ],
    Array [
      Object {
        "type": "text",
        "value": "",
      },
    ],
    Array [
      Object {
        "type": "text",
        "value": "    it('App renders correctly', () => {",
      },
    ],
    Array [
      Object {
        "type": "text",
        "value": "        expect(renderer.create(<App diffText={'diff'} />).toJSON()).toMatchSnapshot();",
      },
    ],
    Array [
      Object {
        "type": "text",
        "value": "    });",
      },
    ],
  ],
  "old": Array [
    Array [
      Object {
        "type": "text",
        "value": "",
      },
    ],
    Array [
      Object {
        "type": "text",
        "value": "",
      },
    ],
    Array [
      Object {
        "type": "text",
        "value": "",
      },
    ],
    Array [
      Object {
        "type": "text",
        "value": "",
      },
    ],
    Array [
      Object {
        "type": "text",
        "value": "",
      },
    ],
    Array [
      Object {
        "type": "text",
        "value": "",
      },
    ],
    Array [
      Object {
        "type": "text",
        "value": "",
      },
    ],
    Array [
      Object {
        "type": "text",
        "value": "",
      },
    ],
    Array [
      Object {
        "type": "text",
        "value": "",
      },
    ],
    Array [
      Object {
        "type": "text",
        "value": "",
      },
    ],
    Array [
      Object {
        "type": "text",
        "value": "",
      },
    ],
    Array [
      Object {
        "type": "text",
        "value": "",
      },
    ],
    Array [
      Object {
        "type": "text",
        "value": "",
      },
    ],
    Array [
      Object {
        "type": "text",
        "value": "",
      },
    ],
    Array [
      Object {
        "type": "text",
        "value": "",
      },
    ],
    Array [
      Object {
        "type": "text",
        "value": "",
      },
    ],
    Array [
      Object {
        "type": "text",
        "value": "",
      },
    ],
    Array [
      Object {
        "type": "text",
        "value": "",
      },
    ],
    Array [
      Object {
        "type": "text",
        "value": "",
      },
    ],
    Array [
      Object {
        "type": "text",
        "value": "",
      },
    ],
    Array [
      Object {
        "type": "text",
        "value": "    it('App renders correctly', () => {",
      },
    ],
    Array [
      Object {
        "type": "text",
        "value": "        expect(renderer.create(<App diffText={'deff'} />).toJSON()).toMatchSnapshot();",
      },
    ],
    Array [
      Object {
        "type": "text",
        "value": "    });",
      },
    ],
  ],
}
`;
